blockchain/chain/src/store/chain_store.rs

```diff
@@ -210,31 +235,49 @@ where
 
         Ok(FullTipset::new(blocks))
     }
-}
+    /// Determines if provided tipset is heavier than existing known heaviest tipset
+    fn update_heaviest(&mut self, ts: &Tipset) -> Result<(), Error> {
+        // TODO determine if expanded tipset is required; see https://github.com/filecoin-project/lotus/blob/testnet/3/chain/store/store.go#L236
+        let new_weight = self.weight(ts)?;
+        let curr_weight = self.weight(&self.heaviest)?;
+
+        if new_weight > curr_weight {
+            // TODO potentially need to deal with re-orgs here
+            info!("New heaviest tipset");
+            self.set_heaviest_tipset(Arc::new(ts.clone()))?;
+        }
+
+        Ok(())
+    }
+
+    /// Returns the weight of provided tipset
+    fn weight(&self, ts: &Tipset) -> Result<BigUint, String> {
+        if ts.is_empty() {
+            return Ok(BigUint::zero());
+        }
+
+        let mut tpow = BigUint::zero();
+        let state = HamtStateTree::new_from_root(self.db.as_ref(), ts.parent_state())?;
+        if let Some(act) = state.get_actor(&*STORAGE_POWER_ACTOR_ADDR)? {
+            if let Some(state) = self
+                .db
+                .get::<PowerState>(&act.state)
+                .map_err(|e| e.to_string())?
+            {
+                tpow = state.total_network_power;
+            }
+        }
+        let log2_p = if tpow > BigUint::zero() {
+            BigUint::from(tpow.bits() - 1)
+        } else {
+            return Err("All power in the net is gone. You network might be disconnected, or the net is dead!".to_owned());
+        };
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use address::Address;
-    use cid::multihash::Identity;
-
-    #[test]
-    fn genesis_test() {
-        let db = db::MemoryDB::default();
-
-        let cs = ChainStore::new(Arc::new(db));
-        let gen_block = BlockHeader::builder()
-            .epoch(1)
-            .weight((2 as u32).into())
-            .messages(Cid::new_from_cbor(&[], Identity))
-            .message_receipts(Cid::new_from_cbor(&[], Identity))
-            .state_root(Cid::new_from_cbor(&[], Identity))
-            .miner_address(Address::new_id(0).unwrap())
-            .build_and_validate()
-            .unwrap();
-
-        assert_eq!(cs.genesis().unwrap(), None);
-        cs.set_genesis(gen_block.clone()).unwrap();
-        assert_eq!(cs.genesis().unwrap(), Some(gen_block));
```

Comment:
Can you explain why exactly this test needs to be removed? Seems like at a high level that function should be able to be called without anything else, why can this not happen?

Reply:
The test needs more setup as it fails during the `weight()` method called from `update_heaviest()`.  I could follow lotus more precisely and have a `put_tipset()` wrapper over `persist_headers()` where `update_heaviest()` would live and have `set_genesis()` only call `persist_headers()` since it only takes in a blockheader anyway. Otherwise, I will refactor the test to account for new required setup.

Reply:
what is the setup that's missing though? that genesis block not existing in the blockstore? seems intuitive that this should not require other setup, but can look more into specifics later

Reply:
No, the setup that's missing is `STORAGE_POWER_ACTOR_ADDR` not being available in the `weight` method. I am adding a `put_tipset` method, similar to lotus; see [here](https://github.com/filecoin-project/lotus/blob/master/chain/store/store.go#L224). As it seems we should not be calling `update_heaviest()` upon initializing the node with genesis.